roomserver/storage/shared/storage.go

```diff
@@ -343,14 +343,52 @@ func (d *Database) GetMembership(ctx context.Context, roomNID types.RoomNID, req
 }
 
 func (d *Database) GetMembershipEventNIDsForRoom(
-	ctx context.Context, roomNID types.RoomNID, joinOnly bool, localOnly bool,
+	ctx context.Context, roomNID types.RoomNID, joinOnly bool, membershipFilter []string, localOnly bool,
 ) ([]types.EventNID, error) {
 	if joinOnly {
 		return d.MembershipTable.SelectMembershipsFromRoomAndMembership(
 			ctx, roomNID, tables.MembershipStateJoin, localOnly,
 		)
 	}
 
+	if len(membershipFilter) > 0 {
+		var eventNIDs []types.EventNID
+		var filteredEventNIDs []types.EventNID
+		var err error
+		leaveOrBan := false
+		for _, filter := range membershipFilter {
+			if filter == "join" {
+				filteredEventNIDs, err = d.MembershipTable.SelectMembershipsFromRoomAndMembership(
+					ctx, roomNID, tables.MembershipStateJoin, localOnly,
+				)
+				if err != nil {
+					return eventNIDs, err
+				}
+				eventNIDs = append(eventNIDs, filteredEventNIDs...)
+			}
+			if filter == "invite" {
+				filteredEventNIDs, err = d.MembershipTable.SelectMembershipsFromRoomAndMembership(
+					ctx, roomNID, tables.MembershipStateInvite, localOnly,
+				)
+				if err != nil {
+					return eventNIDs, err
+				}
+				eventNIDs = append(eventNIDs, filteredEventNIDs...)
+			}
+			if (filter == "ban" || filter == "leave") && !leaveOrBan {
+				filteredEventNIDs, err = d.MembershipTable.SelectMembershipsFromRoomAndMembership(
+					ctx, roomNID, tables.MembershipStateLeaveOrBan, localOnly,
+				)
+				if err != nil {
+					return eventNIDs, err
+				}
+				eventNIDs = append(eventNIDs, filteredEventNIDs...)
```

Comment:
Annoyingly it will be possible for the filter to want bans but not leaves, in which case this will return the wrong data. You need to check this case. In the common case where both leave and ban are present then no filtering needs to be done.